### PR TITLE
Added kernel ridge regression concept entry

### LINKED ISSUE #4859: [Concept Entry] Sklearn kernel-ridge-regression
### Reason/inspiration (optional)

We would like a new entry on the `Kernel Ridge Regression` concept under Sklearn. The entry should go in a new file under `docs/content/sklearn/concepts/kernel-ridge-regression/kernel-ridge-regression.md`.

The entry should include:

- An introduction to the concept
- A `Syntax` section that provides the syntax for the concept (if applicable)
- An `Example` section that provides an example demonstrating the concept in use (if applicable)
- A `Codebyte Example` section that provides a codebyte example demonstrating the concept in use (if applicable)

Please refer to the [concept entry template](https://github.com/Codecademy/docs/blob/main/documentation/concept-entry-template.md), [content standards](https://github.com/Codecademy/docs/blob/main/documentation/content-standards.md), and [markdown style guide](https://curriculum-documentation.codecademy.com/content-guidelines/markdown-style-guide/) when working on the PR for this issue.

### Entry Type (select all that apply)

New Concept Entrty for Existing Topic

### Code of Conduct

- [X] By submitting this issue, I agree to follow Codecademy Doc's [Code of Conduct](https://github.com/Codecademy/docs/blob/main/.github/CODE_OF_CONDUCT.md).

### For Maintainers

- [X] Labels added
- [ ] Issue is assigned

### PR DESCRIPTION
**Pull request by uhck**
### Description

Add kernel-ridge-regression.md concept

### Issue Solved

Closes #4859

### Type of Change

- Adding a new entry

### Checklist

- [x] All writings are my own.
- [x] My entry follows the Codecademy Docs style guide.
- [x] My changes generate no new warnings.
- [x] I have performed a self-review of my own writing and code.
- [x] I have checked my entry and corrected any misspellings.
- [x] I have made corresponding changes to the documentation if needed.
- [x] I have confirmed my changes are not being pushed from my forked `main` branch.
- [x] I have confirmed that I'm pushing from a new branch named after the changes I'm making.
- [x] I have linked any issues that are relevant to this PR in the `Issues Solved` section.

### CODE REVIEW

File: content/sklearn/content/kernel-ridge-regression/kernel-ridge-regression.md [hunk 1]
```diff
@@ -0,0 +1,76 @@
+---
+Title: 'Kernel Ridge Regression'
```
Comment (SaviDahegaonkar) [inline content/sklearn/content/kernel-ridge-regression/kernel-ridge-regression.md:2]:
``` suggestion
Title: 'Kernel Ridge Regression'
```
Title name in docs is always written in lowercase and separated by dashes instead of spaces. Please take a note of it.

**Review by SaviDahegaonkar — CHANGES_REQUESTED** [review]
Hey @uhck,
I had a look at your file and I am finding a lot of corrections that should be made in order to make this PR a valid PR. Please have a look at the [Contribution Guide documentation](https://github.com/Codecademy/docs/tree/main/documentation) and do all the corrections properly. 

Thanks,
Savi

File: content/sklearn/content/kernel-ridge-regression/kernel-ridge-regression.md (hunk 1 above)
Comment (uhck) [reply content/sklearn/content/kernel-ridge-regression/kernel-ridge-regression.md:2]: Do you mean the file name or title name? Not sure why I would need to do that in the title name. This is a concept entry, not a term entry.

Comment (SaviDahegaonkar) [reply content/sklearn/content/kernel-ridge-regression/kernel-ridge-regression.md:2]: > Do you mean the file name or title name? Not sure why I would need to do that in the title name. This is a concept entry, not a term entry.

I mean the title name must be the same as the file name and the title name must always be in lowercase and separated by dashes instead of spaces.

Comment (uhck) [reply content/sklearn/content/kernel-ridge-regression/kernel-ridge-regression.md:2]: Please link me where it states that in the Contribution Guide docs. From what is specified in the style-guide.md file, I followed this template:

**Title: 'The Title'** <-- Not lowercase or separated by dashes.
Description: 'A brief description'
Subjects:
  - 'A subject name'
  - 'A second subject name'
  - 'An nth subject name'
...

Can you please briefly list in bullet points what makes this an invalid PR?

Comment (uhck) [reply content/sklearn/content/kernel-ridge-regression/kernel-ridge-regression.md:2]: Pinging here for a response in what makes this PR invalid.

File: content/sklearn/concepts/kernel-ridge-regression/kernel-ridge-regression.md [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+Description: 'Kernel ridge regression is a sophisticated linear regression model combined with L2 regularization and kernel trick to provide optimal solutions.'
+Subjects:
+  - 'Data Science'
+  - 'Machine Learning'
+Tags:
+  - 'Machine Learning'
+  - 'Python'
+  - 'Scikit-learn'
+CatalogContent:
+  - 'paths/intermediate-machine-learning-skill-path'
```
Comment (SaviDahegaonkar) [inline content/sklearn/concepts/kernel-ridge-regression/kernel-ridge-regression.md:12]:
``` suggestion
  - 'paths/intermediate-machine-learning-skill-path'
```
The course/path mentioned above is not valid. Using the course/path landing page slugs is necessary, If the path you want is not present in the [catalog-content. md file](documentation/catalog-content.md). You can add one.

File: content/sklearn/concepts/kernel-ridge-regression/kernel-ridge-regression.md [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+---
+
+**Kernel ridge regression** is a regression model that combines ridge regression with the kernel trick.
+
+## Ridge Regression
+
+**Ridge regression** is a linear regression model with a least square loss function and L2 regularization. The loss function, equivalent to the sum of the coefficients squared, allows the linear regression model to find the best fit line on the training data as shown in the example below:
```
Comment (SaviDahegaonkar) [inline content/sklearn/concepts/kernel-ridge-regression/kernel-ridge-regression.md:19]:
``` suggestion
**Ridge regression** is a linear regression model with a least square loss function and L2 regularization. The loss function combines the; least square loss with the L2 regularization term and allows the linear regression model to find the best-fit line on the training data. It is shown in the example below:
```
Ridge regression is not just equivalent to the sum of the coefficients squared. It actually combines the least squares loss (which is the sum of the squared differences between the predicted and actual values) with the L2 regularization term.

File: content/sklearn/concepts/kernel-ridge-regression/kernel-ridge-regression.md [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+## Applying the Kernel Trick
+
+However, not all data can be separated linearly. In the real world, data can be randomly distributed, making it difficult to classify the data with linear regression as shown in the example below:
+
+![Kernel trick](https://raw.githubusercontent.com/Codecademy/docs/main/media/kernel-trick.png)
+
+The **kernel trick** allows us to form a more complex model in the original feature space without incurring huge computing costs required to transform the data into a higher dimensional space. It is more efficient and less expensive to use the kernel function, which is the dot product of x-transpose and y in 2-dimensional space, instead of doing more complicated computations in 4-dimensional space. The more dimensions you are working with, the more costly the computations. Because we are mapping data to a higher dimension, there is a greater chance that we may overfit the model and thus, ridge regression balances out the this critical issue by incorporating L2 regularization.
```
Comment (SaviDahegaonkar) [inline content/sklearn/concepts/kernel-ridge-regression/kernel-ridge-regression.md:31]:
``` suggestion
The **kernel trick** allows us to form a more complex model in the original feature space without incurring huge computing costs to transform the data into a higher dimensional space. It is more efficient and less expensive to use the kernel function, which is the dot product of x-transpose and y in 2-dimensional space, instead of doing more complicated computations in 4-dimensional space. The more dimensions you are working with, the more costly the computations. Because we are mapping data to a higher dimension, there is a greater chance that we may overfit the model, and thus, ridge regression balances out this critical issue by incorporating L2 regularization.
```

File: content/sklearn/concepts/kernel-ridge-regression/kernel-ridge-regression.md [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+```pseudo
+# Initialize the classifier
+model = KernelRidge()
+
+# Train the model with training data
+model.fit(x_train, y_train)
+
+# Use the model to predict the outcomes for testing: data
+predictions = model.predict(x_test)
```
Comment (SaviDahegaonkar) [inline content/sklearn/concepts/kernel-ridge-regression/kernel-ridge-regression.md:43]:
```suggestion
```pseudo
# Initialize the classifier
model = KernelRidge()

# Train the model with training data
model.fit(x_train, y_train)

# Use the model to predict the outcomes for testing data
predictions = model.predict(x_test)
```

File: content/sklearn/concepts/kernel-ridge-regression/kernel-ridge-regression.md [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+# Train the model with training data
+krr.fit(X_train, y_train)
+
+# Predict the outcomes for test data
+y_pred = krr.predict(X_test)
+ 
+# Calculate the R2 Score
+r2 = r2_score(y_test, y_pred)
+print(f'R2 Score: {r2}')
+```
```
Comment (SaviDahegaonkar) [inline content/sklearn/concepts/kernel-ridge-regression/kernel-ridge-regression.md:76]:
``` suggestion
```codebyte/python
from sklearn.datasets import load_iris
from sklearn.kernel_ridge import KernelRidge
from sklearn.metrics import r2_score
from sklearn.model_selection import train_test_split

# Load the iris dataset
iris = load_iris()
  
# Store the feature matrix (X) and response vector (y)
X = iris.data
y = iris.target
 
# Split the data set into training and testing data sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)
 
# Initialize the Kernel Ridge Regression classifier
krr = KernelRidge()
 
# Train the model with training data
krr.fit(X_train, y_train)

# Predict the outcomes for test data
y_pred = krr.predict(X_test)
 
# Calculate the R2 Score
r2 = r2_score(y_test, y_pred)
print(f'R2 Score: {r2}')
```
```
However the code is correct but `KernelRidge` is used for regression and not for classification, so using it on the iris dataset is not suitable also the R2 score calculated will be numeric values, and interpreting a numeric value in the context of classification is not appropriate.

**Review by SaviDahegaonkar — CHANGES_REQUESTED** [review]
Hey @uhck,
I have suggested some changes that should be done and also I would like you to know that adding an example would be more helpful and will satisfy the requirements of the [issue ](https://github.com/Codecademy/docs/issues/4859)related to this PR. Please get this done asap.

File: content/sklearn/concepts/kernel-ridge-regression/kernel-ridge-regression.md (hunk 6 above)
Comment (uhck) [reply content/sklearn/concepts/kernel-ridge-regression/kernel-ridge-regression.md:76]: Added an example and reworked the codebyte

**Comment by uhck** [comment]
Thanks for the feedback, updated the commit

File: content/sklearn/concepts/kernel-ridge-regression/kernel-ridge-regression.md [hunk 7]
```diff
@@ -0,0 +1,115 @@
+---
+Title: 'Kernel Ridge Regression'
+Description: 'Kernel ridge regression is a sophisticated linear regression model combined with L2 regularization and kernel trick to handle non-linearities that provide optimal solutions.'
+Subjects:
+  - 'Data Science'
+  - 'Machine Learning'
+Tags:
+  - 'Machine Learning'
+  - 'Python'
+  - 'Scikit-learn'
```
Comment (SaviDahegaonkar) [inline content/sklearn/concepts/kernel-ridge-regression/kernel-ridge-regression.md:10]:
```suggestion
Tags:
  - 'Machine Learning'
  - 'Python'
  - 'Scikit-learn'
```
Include ` Python` in the [tags.md list.](documentation/tags.md)

File: content/sklearn/concepts/kernel-ridge-regression/kernel-ridge-regression.md [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+CatalogContent:
+  - 'paths/data-science'
+  - 'paths/intermediate-machine-learning-skill-path'
```
Comment (SaviDahegaonkar) [inline content/sklearn/concepts/kernel-ridge-regression/kernel-ridge-regression.md:13]:
```suggestion
  - 'paths/intermediate-machine-learning-skill-path'
```
Add 'paths/intermediate-machine-learning-skill-path' in the [catalog-content.md file](documentation/catalog-content.md)

**Review by SaviDahegaonkar — CHANGES_REQUESTED** [review]
Hey @uhck ,
I have suggested a change, make it asap. The entry looks good to me.

Thanks,
Savi

File: content/sklearn/concepts/kernel-ridge-regression/kernel-ridge-regression.md (hunk 8 above)
Comment (uhck) [reply content/sklearn/concepts/kernel-ridge-regression/kernel-ridge-regression.md:13]: Already in there

**Comment by uhck** [comment]
Made requested changes.

**Review by SaviDahegaonkar — APPROVED** [review]
Hey @uhck ,
The entry LGTM! It looks ready for a second round of review.
Thanks,
Savi
